service/src/main/kotlin/fi/espoo/evaka/placement/PlacementTermination.kt

```diff
@@ -0,0 +1,203 @@
+            TerminatablePlacementGroup(
+                type = type,
+                placements = placementsOfSameType,
+                additionalPlacements = additional,
+                startDate = placementsOfSameType.minOf { placement -> placement.startDate },
+                endDate = placementsOfSameType.maxOf { placement -> placement.endDate },
+                unitId = placementsOfSameType[0].unitId,
+                unitName = placementsOfSameType[0].unitName,
+                terminatable = placementsOfSameType[0].terminatable
+            )
+        }
+    }
+    .sortedBy { it.startDate }
+
+private fun ChildPlacement.startsAfter(date: LocalDate): Boolean = this.startDate.isAfter(date)
+
+fun cancelOrTerminatePlacement(
+    tx: Database.Transaction,
+    terminationRequestedDate: LocalDate,
+    placement: ChildPlacement,
+    terminationDate: LocalDate,
+    terminatedBy: AuthenticatedUser.Citizen?,
+) {
+    if (placement.startsAfter(terminationDate)) {
+        tx.cancelPlacement(placement.id)
+    } else {
+        tx.terminatePlacementFrom(
+            terminationRequestedDate,
+            placement.id,
+            terminationDate,
+            terminatedBy?.evakaUserId
+        )
+    }
+}
+
+fun terminateBilledDaycare(
+    tx: Database.Transaction,
+    user: AuthenticatedUser.Citizen,
+    terminationRequestedDate: LocalDate,
+    terminatablePlacementGroup: TerminatablePlacementGroup,
+    terminationDate: LocalDate,
+    childId: ChildId,
+    unitId: DaycareId
+) {
+    // additional placements after termination date are always cancelled
+    terminatablePlacementGroup.additionalPlacements
+        .filter { it.endDate.isAfter(terminationDate) }
+        .forEach { cancelOrTerminatePlacement(tx, terminationRequestedDate, it, terminationDate, user) }
+
+    val preschoolOrPreparatoryWithDaycare = terminatablePlacementGroup.placements
+        .filter {
+            (it.type == PRESCHOOL_DAYCARE || it.type == PREPARATORY_DAYCARE) && it.endDate.isAfter(terminationDate)
+        }
+    // placements which have already been updated on the earlier iteration
+    val placementsToSkip = mutableListOf<ChildPlacement>()
+    for (placement in preschoolOrPreparatoryWithDaycare) {
+        if (placementsToSkip.contains(placement)) {
+            continue
+        }
+        // It is possible that we already have an adjacent placement of the same type, e.g. when the billed DAYCARE part
+        // of a PRESCHOOL_DAYCARE placement has been terminated before. Detect those and prevent duplicates.
+        val adjacentPlacement = terminatablePlacementGroup.placements.find {
+            it.type == when (placement.type) {
+                PRESCHOOL_DAYCARE -> PRESCHOOL
+                PREPARATORY_DAYCARE -> PREPARATORY
+                else -> throw IllegalStateException("Should not be handling any other placement types here")
+            } && it.startDate == placement.endDate.plusDays(1)
```

Comment:
This `when` clause could be extracted into a function that could be used at rows `176` and `192` as well.